daft/execution/shuffles/flight_shuffle/flight_shuffle.py

```diff
@@ -0,0 +1,465 @@
+
+    clients = [flight.FlightClient(address) for address in shuffle_actor_addresses]
+
+    # Generate partitions from the actors
+    async def gen_partitions():
+        loop = asyncio.get_running_loop()
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            queue = asyncio.Queue(maxsize=max_parallel_fetches)
+            fetch_task = asyncio.create_task(fetch_all_partitions(clients, partitions, executor, queue))
+
+            while True:
+                next_item = await queue.get()
+                if next_item is None:
+                    break
+                partition_idx, futures = next_item
+
+                def wait_for_futures(futures):
+                    return [future.result() for future in concurrent.futures.as_completed(futures)]
+
+                # Wait for all the futures to complete, run in threadpool so we don't block the event loop
+                to_reduce = await loop.run_in_executor(
+                    executor,
+                    wait_for_futures,
+                    futures,
+                )
+                yield (partition_idx, to_reduce)
+
+            await fetch_task
+
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
+
+    async_partition_gen = gen_partitions()
+    try:
+        while True:
+            try:
+                partition_idx, to_reduce = loop.run_until_complete(async_partition_gen.__anext__())
+                reduced = MicroPartition.concat(to_reduce)
+                metadata = PartitionMetadata.from_table(reduced)
+
+                yield reduced
+                shuffle_actor_manager.clear_partitions.remote([partition_idx])
+                yield metadata
+
+            except StopAsyncIteration:
+                break
+    finally:
+        # close clients
+        for client in clients:
+            client.close()
+        # close the event loop
+        loop.close()
+
+
+# Figure out which partitions go to which reduce tasks
+# This is a simple round robin assignment
+def optimize_reduce_partition_assignment(actor_node_ids, num_partitions):
+    # Assign 2x number of reduce tasks as there are actors, this seems to work alright in practice
+    num_reduce_tasks = len(actor_node_ids) * 2
+
+    # Map partitions to reduce tasks in round robin fashion
+    partition_to_reduce_task = [partition_idx % num_reduce_tasks for partition_idx in range(num_partitions)]
+
+    # Create the reverse mapping
+    reduce_task_to_partitions = [[] for _ in range(num_reduce_tasks)]
+    for partition_idx, reduce_task_idx in enumerate(partition_to_reduce_task):
+        reduce_task_to_partitions[reduce_task_idx].append(partition_idx)
+
+    return reduce_task_to_partitions, partition_to_reduce_task
+
+
+def run_map_phase(
+    plan: InProgressPhysicalPlan[ray.ObjectRef],
+    map_stage_id: int,
+    shuffle_actor_manager: ShuffleActorManager,
+):
+    # Max number of maps to emit per wave
+    batch_size = next(_ray_num_cpus_provider())
```

Comment:
we likely also want to an an excess amount of tasks. Maybe 2x this?

Reply:
Also we should be refreshing num_cpus inside the loop since this number can change due to autoscaling